### ISSUE TITLE
Rolling update of deployments creates MAX number of pods X 2 according to HPA, regardless of current load

### ISSUE THREAD

**Issue by paalkr** [kubernetes/kubernetes#84142]
**What happened**:
The bug described in https://github.com/kubernetes/kubernetes/issues/72775 is still present in 1.14.8
When a change is made to the running deployment that requires a rolling update to be executed, both the current replicaset and the new replicaset is scaled out to max number of pods regardless of actual load.

**What you expected to happen**:
The bug to be fixed
https://github.com/kubernetes/kubernetes/pull/79035
https://github.com/kubernetes/kubernetes/pull/79708

**How to reproduce it (as minimally and precisely as possible)**:
Follow this guide https://gist.github.com/skaji/575e4a383fac0f0d11b5b51220a6049c
The cluster does NOT have to be an EKS cluster, the bug is not related to the underlying platform or infrastructure.

**Anything else we need to know?**:
This is bug is really critical, and should be addressed and hotfixed as soon as possible.

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:13:54Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"windows/amd64"}                                                                                                                                    Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.8", GitCommit:"211047e9a1922595eaa3a1127ed365e9299a6c23", GitTreeState:"clean", BuildDate:"2019-10-15T12:02:12Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"} 
- Cloud provider or hardware configuration: AWS EC2
- OS (e.g: `cat /etc/os-release`): Container Linux by CoreOS 2247.5.0 (Rhyolite)
- Kernel (e.g. `uname -a`): 4.19.78-coreos
- Install tools: kube-aws v.0.14.2 (https://github.com/kubernetes-incubator/kube-aws)

**Comment by paalkr** [comment]
@kubernetes/sig-autoscalling-bugs

**Comment by paalkr** [comment]
This bug does only apply to the `autoscaling/v2beta2` and `autoscaling/v2beta1` API version, not the `autoscaling/v1` API.

**Comment by thdrnsdk** [comment]
same issue
i think it was really critical iuuse so hope hot fixed asap

**Comment by JoelSpeed** [comment]
I've been looking into this and have made some interesting observations about the behaviour within our systems at least:

Firstly, during the update, we see `New size: x; reason: All metrics below target` even though `x` is greater in size than the current replica count, and this continues to increase over time. This suggests the bug may be somewhere in here: https://github.com/kubernetes/kubernetes/blob/4485c6f18cee9a5d3c3b4e523bd27972b1b53892/pkg/controller/podautoscaler/horizontal.go#L624-L627

`currentReplicas` is set to `scale.Spec.Replicas` and so I believe `desiredReplicas` is being increased (to greater than `currentReplicas`) by the `normalizeDesiredReplicas` method.

Secondly, we see the scale doubling (approximately) during the rolling update. The observation I've made it that it seems to be getting updated to the total count of the old and new replicasets pods combined, as if it is using the number of pods that match the label matcher for the deployment as the desired count when it finally updates the scale. I have yet to find a piece of code that verifies this though, merely an observation.

Hopefully this helps narrow down the bug, it may also be a wild goose chase 😅 I can do more testing if anyone has any ideas, just let me know

**Comment by qoobic** [comment]
Just for information, I am seeing similar behaviour using a custom metric, not CPU - I am using `sum(rate(http_server_requests_seconds_count[5m]))`. In our dev environment the config is `min` of 1, `max` of 3. Even when there are no requests coming in, hpa is scaling to 3 or 4 pods then cooling back down to 3 pods after 5 minutes.

**Comment by shibataka000** [comment]
#### In `Step3`
- CPU Utilization of Pod-1 is `1 / 100 = 0.01`.
- CPU Utilization of Pod-2 is `1 / 100 = 0.01`.

So CPU Utilization of Deployment is `(1 + 1) / (100 + 100) = 0.01`.

```
calculatedReplicas = ceil(readyPodCount * CurrentResourceUtilization / desiredResourceUtilization) = ceil(2 * 0.01 / 0.5) = 1
```

So calculated number of replicas is `1`. But `minReplicas` is `2`, so as a result, new number of replicas is `2`.

#### In `Step 5`
New Pod-3 is created due to RollingUpdate with `maxSurge: 1` and `minUnavailable: 0`.

- CPU Utilization of Pod-1 is `1 / 100 = 0.01`.
- CPU Utilization of Pod-2 is `1 / 100 = 0.01`.
- CPU Utilization of Pod-3 is `100 / 100 = 1.0` because metrics-server can't provide metrics of freshly maded Pod-3 and hpa-controller treats pod's resource utilization as `1.0` when deployment tend to scale in.

So CPU Utilization of Deployment is `(1 + 1 + 100) / (100 + 100 + 100) = 0.34`.

```
ceil(len(Pods) * CurrentResourceUtilization / desiredResourceUtilization) = ceil(3 * 0.34 / 0.5) = 3
```

So calculated number of replicas is `3` in spite of `CurrentResourceUtilization / desiredResourceUtilization < 1.0` .

**Comment by shibataka000** [comment]
I'd like to fix this issue in #85027 .

**Comment by max-rocket-internet** [comment]
@shibataka000 as workaround, can we just set `minUnavailable: 1` to solve it?

**Comment by paalkr** [comment]
@shibataka000 , just to make sure this is not forgotten
Specifying `maxUnavailable: 0` also results in an uncontrolled roll out. Please verify that the fix in https://github.com/kubernetes/kubernetes/pull/85027 also apply to the `maxSurge: 1` and `maxUnavailable: 0` scenario as well.

**Comment by paalkr** [comment]
@max-rocket-internet , if I'm not mistaken this bug is related to `maxUnavailable: 0`, not `minUnavailable: 0`. So specifying `minUnavailable: 1` won't help.

**Comment by max-rocket-internet** [comment]
Sorry, I meant `maxUnavailable: 0` but copy pasted the incorrect text. I think `minUnavailable` is simply a typo.

So, as workaround, can we just set `maxUnavailable: 1` to solve it?

**Comment by paalkr** [comment]
`maxUnavailable: 1` is really not a good workaround. Say you have a deployment with 200 pods, `maxUnavailable: 1` would slow down the roll out dramatically, as only one new pod could be created at a time, instead of all 200.

If the startup time for your pod is 10 seconds, replacing all 200 pods would take 10 seconds with `maxUnavailable: 0` and 2000 seconds (33.3 minutes) with `maxUnavailable: 1`

**Comment by shibataka000** [comment]
@paalkr  @max-rocket-internet I'm sorry. `minUnavailable` is typo. `maxUnavailable` is correct.

**Comment by shibataka000** [comment]
@max-rocket-internet 

> as workaround, can we just set maxUnavailable: 1 to solve it?

Unfortunately no. When I try scenario described in https://gist.github.com/skaji/575e4a383fac0f0d11b5b51220a6049c with `maxUnavailable: 1`, HPA scale deployment regardless of current load. `len(metrics)` become `3` depends on timing.

**Comment by paalkr** [comment]
What is the status of this issue? I still believe it's a critical bug that needs to be addressed.

**Comment by caiohasouza** [comment]
+1

**Comment by danielgblanco** [comment]
We may be running into the same or a problem that may be connected to this. We run a deployment with

```
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 0
      maxSurge: 1
```
The HPA, which targets the deployment, is set to scale on a custom metric which is scraped every 30 seconds (as the value is a calculated as a `rate` it takes 60 seconds for the first value to appear).

When the rolling update runs, even with completely idle deployment at 1 replicas, the HPA starts to add replicas to the deployment until it either hits the max replicas (which is 8 for this deployment) or until the first Pod generates a metric. The in [calcPlainMetricReplicas](https://github.com/kubernetes/kubernetes/blob/0a2a69add2c914675898bd9018010a03eca8f693/pkg/controller/podautoscaler/replica_calculator.go#L168) method works as intended. It uses 100% of the target value for pods missing metrics, so, with 2 replicas running, one at 100% and the other one at 1% it still returns 2 replicas as the desired replicas count.

I believe the problem is related to how the current number of replicas is taken from the deployment spec during HPA reconciliation ([code](https://github.com/kubernetes/kubernetes/blob/0a2a69add2c914675898bd9018010a03eca8f693/pkg/controller/podautoscaler/horizontal.go#L582)). During a rolling update, we saw the deployment spec and the deployment status being out of sync for long periods of time (over 3 minutes).

```
$ while sleep 1; do
>     date
>     kubectl get deployment my-deployment -o=jsonpath='{.spec.replicas} {.status.replicas}'
>     echo
> done

Mon  2 Mar 2020 14:42:26 GMT
1 1
[...]
Mon  2 Mar 2020 14:43:22 GMT
1 1
Mon  2 Mar 2020 14:43:23 GMT
1 2
[...]
Mon  2 Mar 2020 14:43:28 GMT
2 3
[...]
Mon  2 Mar 2020 14:43:42 GMT
3 4
[...]
Mon  2 Mar 2020 14:43:57 GMT
4 5
[...]
Mon  2 Mar 2020 14:44:13 GMT
5 6
[...]
Mon  2 Mar 2020 14:44:27 GMT
6 7
[...]
Mon  2 Mar 2020 14:47:50 GMT
6 6
[...]
Mon  2 Mar 2020 14:49:28 GMT
5 5
[...]
Mon  2 Mar 2020 14:49:43 GMT
4 4
[...]
Mon  2 Mar 2020 14:49:58 GMT
3 3
[...]
Mon  2 Mar 2020 14:52:28 GMT
2 2
[...]
Mon  2 Mar 2020 14:52:58 GMT
1 1
```
This does not always happen, but it happens most of the times we do a rolling update. The HPA status is also confusing, as it indicates that the metric is above target when in fact is well below target:

```
  Type    Reason             Age                 From                       Message
  ----    ------             ----                ----                       -------
  Normal  SuccessfulRescale  2m44s (x3 over 5d)  horizontal-pod-autoscaler  New size: 2; reason: pods metric my_custom_metric above target
  Normal  SuccessfulRescale  2m29s (x3 over 5d)  horizontal-pod-autoscaler  New size: 3; reason: pods metric my_custom_metric above target
  Normal  SuccessfulRescale  2m14s (x3 over 5d)  horizontal-pod-autoscaler  New size: 4; reason: pods metric my_custom_metric above target
  Normal  SuccessfulRescale  119s (x3 over 5d)   horizontal-pod-autoscaler  New size: 5; reason: pods metric my_custom_metric above target
  Normal  SuccessfulRescale  104s                horizontal-pod-autoscaler  New size: 6; reason: pods metric my_custom_metric above target
```

Should the HPA reconciliation use the current number of replicas from the status rather that from the spec?

**Comment by josephburnett** [comment]
@paalkr thanks for the report.  I am able to reproduce this on 1.14.10 with your test yaml.

I will find the root cause and come back to you shortly.

**Comment by bgpat** [comment]
This issue is reproduced in my environment (v1.15.9) too, and it's resolved by replacing the metric type from [PodsMetricSource](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#podsmetricsource-v2beta1-autoscaling) to [ObjectMetricSource](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#objectmetricsource-v2beta1-autoscaling). I suspect the formula calculating the average is wrong.

**Comment by josephburnett** [comment]
I've cherry picked #85027 on top of 1.14.8 and it indeed fixes the bug.  I've approved the PR and requested to fix it in `calcPlainMetricReplicas` as well.

**Comment by paalkr** [comment]
@josephburnett , thx a lot! So in the next release of 1.14, 1.15, 1.16 and 1.17 this bug will be fixed?

**Comment by gjtempleton** [comment]
Unless I'm mistaken this won't be cherry-picked back to 1.14 as it's fallen out of support as of the GA release of 1.17 as per [the support policy](https://kubernetes.io/docs/setup/release/version-skew-policy/#supported-versions).

At best it'll be getting back to 1.15.

**Comment by josephburnett** [comment]
I'll work on backporting this to 1.15 - 1.18.

**Comment by josephburnett** [comment]
Cherry pick process (for reference): https://github.com/kubernetes/community/blob/83af19fd21fe3af7cd88630e7239b974091645d2/contributors/devel/sig-release/cherry-picks.md#cherry-pick-review

1:15: https://github.com/kubernetes/kubernetes/pull/89514
1.16: https://github.com/kubernetes/kubernetes/pull/89515
1.17: https://github.com/kubernetes/kubernetes/pull/89516
1.18: https://github.com/kubernetes/kubernetes/pull/89517

**Comment by shyamjos** [comment]
This bug is affecting us badly , Is this fix merged in 1.16.x release ? if not is there a temporary solution to fix this issue

**Comment by djjayeeta** [comment]
@shyamjos 
This fix is available from 1.16.9.
